app/workers/follows/update_points_worker.rb

```diff
@@ -31,8 +31,8 @@ def calculate_implicit_points(tag, user)
         .pluck(:reactable_id).last(100)
       last_100_long_page_view_article_ids = user.page_views.where(time_tracked_in_seconds: 45..)
         .pluck(:article_id).last(100)
-      articles = Article.where(id: last_100_reactable_ids + last_100_long_page_view_article_ids)
-      tags = articles.pluck(:cached_tag_list).compact.flat_map { |list| list.split(", ") }
+      articles = Article.where(id: last_100_reactable_ids + last_100_long_page_view_article_ids).joins(:tags)
```

Comment:
okay, so I said "behavior neutral", but here we reference a joined association and use it when we access the tag_list of each article. This is less _efficient_ than using the cached_tag_list string, but more reliable.

It's tradeoffs all the way down.

![image](https://user-images.githubusercontent.com/1237369/144327364-f4367813-d77b-4a3d-a3cf-9350958c11d1.png)

Reply:
Is there a way to test how badly this will impact performance before we deploy it, maybe a test run via a Rails console on a subset of data or something?

Reply:
Local testing with logs disabled over random users and random article ids gives about 80-85ms to run on main, and 115-125ms to run on the branch, and about 170ms when the tags join, removed in 0d25b96, was still in place. Honeycomb shows a slightly higher duration (since it's timing the sidekiq worker process and any involved overhead, and the latency for the database will be higher than local testing, and my data is probably not a realistic representative), with about 7000 jobs processed in the last day, and 47000 in the past week, the durations not usually out of the 100-250ms range. 

We can definitely compare post-release in honeycomb by average or p95 on duration, or I can get a test plan in place to test directly (patch the class and call it).

I'm not 100% sure this is idempotent (so I'm not sure I want to run this in production), this seems to be a consequence of the `ORDER BY RANDOM()` sort for tag followers during the `adjust_other_tag_follows_of_user` and the fact that scoring a follow once causes updates to other follow scores (so while the calculation for a follow usually results in the same result from finalized_points, the user wide adjustment downward occurs every time the perform method is called - rather than exactly once per follow - in the unusual case of retried jobs.

Definitely running this more than once on the same user/article argument pair results in an unintended skew on the other follows - this is especially true when a user has few follows since the same set gets sampled repeatedly

The article I was testing against had 2 of the tags the user follows (3 and 13) and penalized repeatedly the other tags:

```
user.follows.pluck(:id, :points, :followable_id)                                                                                                       
=> [[271, 0.000544628241296005, 7],
 [495, 0.0004633495687483912, 1],
 [592, 0.0016855283136629118, 12],
 [2392, 2.9459101490553135, 3],
 [2393, 0.0015160115155284258, 8],
 [1786, 0.0009879738981274629, 4],
 [1997, 0.0022633871512712533, 9],
 [492, 3.3978952727983707, 13]]
```

Interestingly - this same behavior is replicable from the front end, by finding an article tagged with some of the tags you follow - then reacting on and off (heart, heart off, heart, heart off, and so on).

```
# initial state looks like this
[341] pry(main)> User.find_by(email: 'admin@forem.local').follows.pluck(:points, :followable_id)
=> [[2.791759469228055, 3],
 [2.9459101490553135, 7],
 [3.3299373673424033, 8],
 [2.8983457434967277, 6],
 [2.735924279843494, 9],
 [3.3978952727983707, 13]]

# repeatedly heart and unheart (or unicorns, doesn't matter) an article with tags 3 and 13
=> [[0.7171629047427074, 8],
 [0.5757530490858022, 6],
 [0.6388244407392527, 9],
 [3.6390573296152584, 13],
 [0.27445354472714856, 7],
 [2.9459101490553135, 3]]

# find another article with a followed tag (9 and 7 in this case) and add a heart one time to it
# this is because each of the articles tags are given a calculated score (this part _is_ idempotent)
=> [[3717, 0.5296743630063926, 8],
 [3718, 0.4083941696614332, 6],
 [3719, 2.9459101490553135, 9],
 [3744, 3.6390573296152584, 13],
 [3741, 3.0794415416798357, 7],
 [3742, 2.886991946074207, 3]]
```

I don't think in practice this is a huge problem - but it looks like while following a tag makes it eligible to show in the feed, not liking posts with that tag trains the feed to show less of them (damping the follow) - this seems like it was a design constraint - but unliking a post doesn't have any restorative function (points before like != points after like and unlike) and running this worker repeatedly appears to introduce some change in the state for users... this tag points appears to be used by the large forem experimental via the ArticleScoreCalculatorForUser, as close as I can tell the WeightedQueryStrategy does not use the score (except insomuch as it's non-negative).

Reply:
all this to say - I think this introduces a roughly 30% performance penalty on a job we run once every 10s or so on DEV, and shouldn't be a problem either way.

Since I merged the work-around test change this is a low-value change and if it's high risk we can discard it.

Reply:
Thanks — as always — for the detail @djuber, I'm ok with whatever you think is best in this case.